staging/src/k8s.io/pod-security-admission/admission/attributes.go

```diff
@@ -0,0 +1,156 @@
+	GetSubresource() string
+	// GetOperation is the operation being performed
+	GetOperation() admissionv1.Operation
+
+	// GetObject returns the parsed v1 Object from incoming request.
+	GetObject() (runtime.Object, error)
+	// GetOldObject returns the parsed v1 existing object. Only populated for UPDATE requests.
+	GetOldObject() (runtime.Object, error)
+	// GetUserName is the requesting user's authenticated name.
+	GetUserName() string
+}
+
+// AttributesRecord is a simple struct implementing the Attributes interface.
+type AttributesRecord struct {
+	Namespace   string
+	Resource    schema.GroupVersionResource
+	Subresource string
+	Operation   admissionv1.Operation
+	Object      runtime.Object
+	OldObject   runtime.Object
+	Username    string
+}
+
+func (a *AttributesRecord) GetNamespace() string {
+	return a.Namespace
+}
+func (a *AttributesRecord) GetResource() schema.GroupVersionResource {
+	return a.Resource
+}
+func (a *AttributesRecord) GetSubresource() string {
+	return a.Subresource
+}
+func (a *AttributesRecord) GetOperation() admissionv1.Operation {
+	return a.Operation
+}
+func (a *AttributesRecord) GetUserName() string {
+	return a.Username
+}
+func (a *AttributesRecord) GetObject() (runtime.Object, error) {
+	return a.Object, nil
+}
+func (a *AttributesRecord) GetOldObject() (runtime.Object, error) {
+	return a.OldObject, nil
+}
+
+// Decoder is an interface for decoding serialized objects from an admission request.
+type Decoder interface {
+	// DecodeRaw decodes a RawExtension object into the passed-in runtime.Object.
+	DecodeRaw(rawObj runtime.RawExtension, into runtime.Object) error
+}
+
+// RequestAttributes adapts an admission.Request to the Attributes interface.
+func RequestAttributes(request admissionv1.AdmissionRequest, decoder Decoder) Attributes {
+	return &attributes{
+		r:       request,
+		decoder: decoder,
+	}
+}
+
+// attributes is an interface used by AdmissionController to get information about a request
+// that is used to make an admission decision.
+type attributes struct {
+	r       admissionv1.AdmissionRequest
+	decoder Decoder
+}
+
+func (a *attributes) GetNamespace() string {
+	return a.r.Namespace
+}
+func (a *attributes) GetResource() schema.GroupVersionResource {
+	return schema.GroupVersionResource(*a.r.RequestResource)
+}
+func (a *attributes) GetSubresource() string {
+	return a.r.RequestSubResource
+}
+func (a *attributes) GetOperation() admissionv1.Operation {
+	return a.r.Operation
+}
+func (a *attributes) GetUserName() string {
+	return a.r.UserInfo.Username
+}
+func (a *attributes) GetObject() (runtime.Object, error) {
+	return a.decode(a.r.Object)
+}
+func (a *attributes) GetOldObject() (runtime.Object, error) {
+	return a.decode(a.r.OldObject)
+}
+func (a *attributes) decode(in runtime.RawExtension) (runtime.Object, error) {
```

Comment:
this bit is interesting… someone who wanted to run the admission hook on their own types would need to do a custom PodSpecExtractor and custome decoder? that bit feels like it should be paired into the same object or interface… not sure

Reply:
Interesting idea. This one also needs to know about name spaces, but I imagine if someone wanted to do their own type, they would probably just wrap (or copy) the built in.

Reply:
this is already taking a Decoder interface... I more meant this type switching should live in that interface:

`DecodeRaw(groupResource schema.GroupResource, rawObj runtime.RawExtension) (runtime.Object, error)`

and there be a default decoder impl that understands namespaces/pods/built-in types

Reply:
I realized I could eliminate most of this and just use the generic runtime.Decoder interface. Much improved.